timesketch/lib/analyzers/safebrowsing.py

```diff
@@ -0,0 +1,250 @@
+            urls: URLs
+            platforms: platformTypes field of threatInfo
+            types: threatTypes field of threatInfo
+        Returns:
+            Dict of URLs with the hits
+        """
+        results = {}
+
+        endpoint = 'https://safebrowsing.googleapis.com/v4/threatMatches:find'
+        api_client = {
+            'clientId': self._google_client_id,
+            'clientVersion': self._google_client_version,
+        }
+
+        for i in range(0, len(urls), self._SAFE_BROWSING_BULK_LIMIT):
+            body = {
+                'client': api_client,
+                'threatInfo': {
+                    'platformTypes': platforms,
+                    'threatTypes': types,
+                    'threatEntryTypes': ['URL'],
+                    'threatEntries': [
+                        {'url': url}
+                        for url in urls[i:i + self._SAFE_BROWSING_BULK_LIMIT]
+                    ],
+                },
+            }
+
+            r = requests.post(
+                endpoint,
+                params={'key': self._safebrowsing_api_key},
+                json=body,
+            )
+
+            r.raise_for_status()
+
+            result = r.json()
+            if result and 'matches' in result:
+                for match in result['matches']:
+                    threat_result = match.copy()
+
+                    # Keeping only interesting keys
+                    for key in match.keys():
+                        if key not in self._SAFEBROWSING_ENTRY_KEEP:
+                            threat_result.pop(key)
+
+                    results[match['threat']['url']] = threat_result
+
+        return results
+
+    def _sanitize_url(self, url_entry):
+        """Finds http[s]:// in 'url_entry' and returns its content from there.
+
+        Args:
+            url_entry: a URL, with some other characters before and after
+
+        Returns:
+            String with the URL
+        """
+        m = self._URL_BEGINNING_RE.search(url_entry)
+
+        if m:
+            return m.group(1)
+
+        return None
+
+    def run(self):
+        """Entry point for the analyzer.
+
+        Returns:
+            String with summary of the analyzer result
+        """
+        query = (
+            '{"query": { "bool": { "should": [ '
+            '{ "exists" : { "field" : "url" }} ] } } }')
+
+        return_fields = ['url']
+
+        events = self.event_stream(
+            query_dsl=query,
+            return_fields=return_fields,
+        )
+
+        urls = {}
+
+        for event in events:
+            url = self._sanitize_url(event.source.get('url'))
+
+            if not url:
+                continue
+
+            urls.setdefault(url, []).append(event)
+
+        # Exit early if there are no URLs in the data set to analyze.
+        if not urls:
+            return 'No URLs to analyze.'
+
+        if not self._safebrowsing_api_key:
+            return 'Safe Browsing API requires an API key!'
```

Comment:
I would move this up much higher... no need to do even query TS or check anything first... this should be the first check